File: src/prediction/price_interval.py
"""Price interval estimation for production laptop price prediction."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


PROJECT_ROOT = Path(__file__).resolve().parents[2]
DEFAULT_CONFIG_PATH = PROJECT_ROOT / "models" / "price_interval_config.json"


def load_interval_config(config_path: str | Path | None = None) -> dict[str, Any]:
    path = Path(config_path) if config_path is not None else DEFAULT_CONFIG_PATH
    if not path.exists():
        raise FileNotFoundError(f"Price interval config not found at: {path}")
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


_CONFIG = load_interval_config()

RAW_LAPTOP_FIELDS = list(_CONFIG["raw_laptop_fields"])
CRITICAL_FIELDS = tuple(_CONFIG["critical_fields"])
SEGMENT_RMSE: dict[str, float] = dict(_CONFIG["segment_rmse"])
SEGMENT_UPPER_BOUNDS = tuple(_CONFIG["segment_upper_bounds"])
MISSING_FIELD_WEIGHTS: dict[str, float] = dict(_CONFIG["missing_field_weights"])
NO_INFO_FLAG_WEIGHT = float(_CONFIG["no_info_flag_weight"])
MAX_UNCERTAINTY_MULTIPLIER = float(_CONFIG["max_uncertainty_multiplier"])
MIN_PRICE_MILLION = float(_CONFIG["min_price_million"])
API_VERSION = int(_CONFIG.get("api_version", 2))


def _is_missing(value: Any) -> bool:
    if value is None:
        return True
    if isinstance(value, str):
        return not value.strip()
    if isinstance(value, float) and value != value:
        return True
    return False


def price_segment(predicted_price_million: float) -> str:
    """Map a point prediction to the holdout price segment used for error calibration."""
    price = max(float(predicted_price_million), 0.0)
    if price <= SEGMENT_UPPER_BOUNDS[0]:
        return "Q1_low"
    if price <= SEGMENT_UPPER_BOUNDS[1]:
        return "Q2"
    if price <= SEGMENT_UPPER_BOUNDS[2]:
        return "Q3"
    if price <= SEGMENT_UPPER_BOUNDS[3]:
        return "Q4"
    return "Q5_high"


def missing_fields(raw_features: dict[str, Any]) -> list[str]:
    return [field for field in RAW_LAPTOP_FIELDS if _is_missing(raw_features.get(field))]


def completeness_pct(raw_features: dict[str, Any]) -> float:
    present = len(RAW_LAPTOP_FIELDS) - len(missing_fields(raw_features))
    return round(100.0 * present / len(RAW_LAPTOP_FIELDS), 1)


def uncertainty_multiplier(
    raw_features: dict[str, Any],
    encoded_features: dict[str, float | int] | None = None,
) -> float:
    multiplier = 1.0
    for field, weight in MISSING_FIELD_WEIGHTS.items():
        if _is_missing(raw_features.get(field)):
            multiplier += weight

    if encoded_features:
        no_info_flags = sum(
            1
            for key, value in encoded_features.items()
            if key.startswith("no_info_") and float(value) == 1.0
        )
        multiplier += no_info_flags * NO_INFO_FLAG_WEIGHT

    return min(multiplier, MAX_UNCERTAINTY_MULTIPLIER)


def uncertainty_level(multiplier: float, completeness: float) -> str:
    if multiplier <= 1.2 and completeness >= 80:
        return "low"
    if multiplier <= 1.7 and completeness >= 55:
        return "medium"
    return "high"


def build_price_prediction(
    predicted_price_million: float,
    raw_features: dict[str, Any],
    encoded_features: dict[str, float | int] | None = None,
) -> dict[str, Any]:
    """Return point estimate plus a calibrated price interval for Problem 1."""
    point = round(float(predicted_price_million), 3)
    segment = price_segment(point)
    base_margin = SEGMENT_RMSE[segment]
    multiplier = uncertainty_multiplier(raw_features, encoded_features)
    half_width = base_margin * multiplier
    low = max(MIN_PRICE_MILLION, point - half_width)
    high = point + half_width
    completeness = completeness_pct(raw_features)
    missing = missing_fields(raw_features)
    level = uncertainty_level(multiplier, completeness)

    return {
        "predicted_price": point,
        "price_unit": "million_vnd",
        "price_range": {
            "low": round(low, 2),
            "high": round(high, 2),
            "half_width_million_vnd": round(half_width, 2),
        },
        "price_segment": segment,
        "input_completeness_pct": completeness,
        "missing_fields": missing,
        "uncertainty": {
            "level": level,
            "multiplier": round(multiplier, 2),
            "base_rmse_million_vnd": base_margin,
            "reason": _uncertainty_reason(level, missing),
        },
    }


def _uncertainty_reason(level: str, missing: list[str]) -> str:
    if not missing:
        return "Đủ thông tin chính; khoảng giá dựa trên sai số holdout theo phân khúc giá."

    critical_missing = [field for field in missing if field in CRITICAL_FIELDS]
    if level == "high":
        if critical_missing:
            return (
                "Thiếu nhiều thông tin quan trọng "
                f"({', '.join(critical_missing[:4])}); khoảng giá được nới rộng."
            )
        return "Một số thông tin còn thiếu; khoảng giá được nới rộng để phản ánh độ không chắc chắn."

    if critical_missing:
        return (
            f"Thiếu {', '.join(critical_missing[:3])}; "
            "khoảng giá rộng hơn mức trung bình."
        )
    return "Thiếu vài thông tin phụ; khoảng giá điều chỉnh nhẹ theo mức độ hoàn thiện input."

File: app/frontend/price-interval.js
(function initPriceInterval(global) {
  const DEFAULT_CONFIG = {
    raw_laptop_fields: [
      "brand",
      "model",
      "ram_gb",
      "storage_gb",
      "storage_type",
      "screen_size_inch",
      "cpu_text",
      "cpu_brand",
      "cpu_family",
      "cpu_generation",
      "cpu_suffix",
      "gpu_text",
      "condition",
      "warranty_status",
    ],
    critical_fields: ["brand", "model", "ram_gb", "storage_gb", "cpu_text", "gpu_text"],
    segment_rmse: {
      Q1_low: 2.805,
      Q2: 2.575,
      Q3: 3.927,
      Q4: 4.234,
      Q5_high: 11.268,
    },
    segment_upper_bounds: [4.725, 8.407, 13.576, 27.362],
    missing_field_weights: {
      brand: 0.18,
      model: 0.14,
      ram_gb: 0.16,
      storage_gb: 0.14,
      cpu_text: 0.16,
      gpu_text: 0.12,
      storage_type: 0.06,
      screen_size_inch: 0.05,
      condition: 0.04,
      warranty_status: 0.03,
      cpu_brand: 0.03,
      cpu_family: 0.03,
      cpu_generation: 0.02,
      cpu_suffix: 0.02,
    },
    no_info_flag_weight: 0.04,
    max_uncertainty_multiplier: 2.6,
    min_price_million: 0.5,
  };

  let config = DEFAULT_CONFIG;

  function setConfig(nextConfig) {
    if (nextConfig && typeof nextConfig === "object") {
      config = { ...DEFAULT_CONFIG, ...nextConfig };
    }
  }

  function isMissing(value) {
    if (value === null || value === undefined) return true;
    if (typeof value === "string") return !value.trim();
    if (typeof value === "number" && Number.isNaN(value)) return true;
    return false;
  }

  function priceSegment(predictedPriceMillion) {
    const price = Math.max(Number(predictedPriceMillion) || 0, 0);
    const bounds = config.segment_upper_bounds;
    if (price <= bounds[0]) return "Q1_low";
    if (price <= bounds[1]) return "Q2";
    if (price <= bounds[2]) return "Q3";
    if (price <= bounds[3]) return "Q4";
    return "Q5_high";
  }

  function missingFields(rawFeatures) {
    return config.raw_laptop_fields.filter((field) => isMissing(rawFeatures?.[field]));
  }

  function completenessPct(rawFeatures) {
    const missingCount = missingFields(rawFeatures).length;
    const present = config.raw_laptop_fields.length - missingCount;
    return Math.round((100 * present) / config.raw_laptop_fields.length * 10) / 10;
  }

  function uncertaintyMultiplier(rawFeatures, encodedFeatures) {
    let multiplier = 1;
    for (const [field, weight] of Object.entries(config.missing_field_weights)) {
      if (isMissing(rawFeatures?.[field])) multiplier += weight;
    }

    if (encodedFeatures) {
      const noInfoFlags = Object.entries(encodedFeatures).filter(
        ([key, value]) => key.startsWith("no_info_") && Number(value) === 1,
      ).length;
      multiplier += noInfoFlags * config.no_info_flag_weight;
    }

    return Math.min(multiplier, config.max_uncertainty_multiplier);
  }

  function uncertaintyLevel(multiplier, completeness) {
    if (multiplier <= 1.2 && completeness >= 80) return "low";
    if (multiplier <= 1.7 && completeness >= 55) return "medium";
    return "high";
  }

  function uncertaintyReason(level, missing) {
    if (!missing.length) {
      return "Đủ thông tin chính; khoảng giá dựa trên độ chính xác theo phân khúc giá.";
    }

    const criticalMissing = missing.filter((field) => config.critical_fields.includes(field));
    if (level === "high") {
      if (criticalMissing.length) {
        return `Thiếu nhiều thông tin quan trọng (${criticalMissing.slice(0, 4).join(", ")}); khoảng giá được nới rộng.`;
      }
      return "Một số thông tin còn thiếu; khoảng giá được nới rộng để phản ánh độ không chắc chắn.";
    }

    if (criticalMissing.length) {
      return `Thiếu ${criticalMissing.slice(0, 3).join(", ")}; khoảng giá rộng hơn mức trung bình.`;
    }
    return "Thiếu vài thông tin phụ; khoảng giá điều chỉnh nhẹ theo mức độ hoàn thiện input.";
  }

  function buildPricePrediction(predictedPriceMillion, rawFeatures, encodedFeatures) {
    const point = Math.round(Number(predictedPriceMillion) * 1000) / 1000;
    const segment = priceSegment(point);
    const baseMargin = config.segment_rmse[segment];
    const multiplier = uncertaintyMultiplier(rawFeatures || {}, encodedFeatures || null);
    const halfWidth = baseMargin * multiplier;
    const low = Math.max(config.min_price_million, point - halfWidth);
    const high = point + halfWidth;
    const completeness = completenessPct(rawFeatures || {});
    const missing = missingFields(rawFeatures || {});
    const level = uncertaintyLevel(multiplier, completeness);

    return {
      predicted_price: point,
      price_unit: "million_vnd",
      price_range: {
        low: Math.round(low * 100) / 100,
        high: Math.round(high * 100) / 100,
        half_width_million_vnd: Math.round(halfWidth * 100) / 100,
      },
      price_segment: segment,
      input_completeness_pct: completeness,
      missing_fields: missing,
      uncertainty: {
        level,
        multiplier: Math.round(multiplier * 100) / 100,
        base_rmse_million_vnd: baseMargin,
        reason: uncertaintyReason(level, missing),
      },
    };
  }

  function hasCompleteRange(payload) {
    const low = Number(payload?.price_range?.low);
    const high = Number(payload?.price_range?.high);
    return Number.isFinite(low) && Number.isFinite(high);
  }

  function enrichPrediction(payload) {
    const point = Number(payload?.predicted_price);
    if (!Number.isFinite(point)) return payload;

    if (
      hasCompleteRange(payload) &&
      payload.input_completeness_pct != null &&
      payload.uncertainty?.level
    ) {
      return payload;
    }

    const enriched = buildPricePrediction(
      point,
      payload.raw_features || {},
      payload.encoded_features || null,
    );

    return {
      ...payload,
      ...enriched,
      range_source: hasCompleteRange(payload) ? "server" : "client_fallback",
    };
  }

  global.PriceInterval = {
    setConfig,
    buildPricePrediction,
    enrichPrediction,
    hasCompleteRange,
  };
})(window);
